### PR TITLE
Cyberark Audit service integration

### PR DESCRIPTION
**Pull request by nitsan-tzur**
Required items, please complete
   
Change(s):
Add new Data Connector for CyberArk Audit service.

Checked that the validations are passing and have addressed any issues that are present:
arm-ttk validation passed

### CODE REVIEW

**Comment by nitsan-tzur** [comment]
@microsoft-github-policy-service agree

**Comment by v-prasadboke** [comment]
Hello @nitsan-tzur, 
Just to confirm. You used V3 tool to package the solution?

Python libraries mentioned in requirements.txt doesnt have specific version. Please mention version as well.

**Comment by nitsan-tzur** [comment]
> Hello @nitsan-tzur, Just to confirm. You used V3 tool to package the solution?
> 
> Python libraries mentioned in requirements.txt doesnt have specific version. Please mention version as well.
Hi, 
I did use the V3 tool. Added concrete versions to requirements.txt

File: Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConnector/exporter.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    def _gen_chunks_to_object(self, data: list, chunk_size: int):
+        chunk = []
+        for index, line in enumerate(data):
+            if index % chunk_size == 0 and index > 0:
+                yield chunk
+                del chunk[:]
+            chunk.append(line)
+        yield chunk
+
+    def post_data(self, data: list) -> int:
```
[the remark this answers is not in the document]
Comment (nitsan-tzur) [reply Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConnector/exporter.py]: fixed

File: Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConnector/exporter.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
+import base64
+import datetime
+import hashlib
+import hmac
+import json
+import logging
+import requests
+import os
+
+from azure.core.exceptions import HttpResponseError
```
[the remark this answers is not in the document]
Comment (nitsan-tzur) [reply Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConnector/exporter.py]: fixed

File: Solutions/CyberArkAudit/Data Connectors/CyberArkAudit_API_FunctionApp.json
Comment (v-prasadboke) [inline Solutions/CyberArkAudit/Data Connectors/CyberArkAudit_API_FunctionApp.json:1]: Please add connectivity criteria to the data connector

File: Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConn.zip
Comment (v-prasadboke) [inline Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConn.zip:1]: The zip files structure is incorrect and zip is missing with python libraries. 
Please correct  the following.

You can refer to this Data connectors zip for more clarification 
Solutions/BitSight/Data Connectors/BitSightDataConnector/BitSight.zip

Comment (nitsan-tzur) [reply Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConn.zip:1]: I compared my CyberArkAuditSentinelConn.zip with the suggested BitSight.zip file
Both contain:
- Modules will all Python files included
- requirements.txt, host.json and proxies.josn on top level folder
image
Can you please point out what is missing or what should be different in the structure?

**Comment by v-prasadboke** [comment]
Please create a custom table named as **CyberArk_AuditEvents_CL** at location .script/tests/KqlvalidationsTests/CustomTables

File: Solutions/CyberArkAudit/Data Connectors/CyberArkAudit_API_FunctionApp.json
Comment (nitsan-tzur) [reply Solutions/CyberArkAudit/Data Connectors/CyberArkAudit_API_FunctionApp.json:1]: fixed

File: Solutions/CyberArkAudit/SolutionMetadata.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+	"publisherId": "cyberark",
+	"offerId": "cyberark_audit_sentinel",
+	"firstPublishDate": "2024-03-01",
+	"lastPublishedDate": "2023-04-01",
+	"providers": ["CyberArk"],
+	"categories": {
+		"domains" : ["Security - Log Analytics", "SIEM"],
```
[the remark this answers is not in the document]
Comment (nitsan-tzur) [reply Solutions/CyberArkAudit/SolutionMetadata.json]: fixed

File: Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConn.zip
Comment (nitsan-tzur) [reply Solutions/CyberArkAudit/Data Connectors/CyberArkAuditSentinelConn.zip:1]: add .python_packages

**Comment by nitsan-tzur** [comment]
> Please create a custom table named as **CyberArk_AuditEvents_CL** at location .script/tests/KqlvalidationsTests/CustomTables
> 
> Theres an extra folder for .python packages which includes python libraries mentioned in requirements.txt image

done

**Comment by v-prasadboke** [comment]
Hello @nitsan-tzur, I've packaged the solution. Can you provide me credentials to test the content of the solution. 
you can mail me them at v-prasadboke@microsoft.com

**Comment by nitsan-tzur** [comment]
> Hello @nitsan-tzur, I've packaged the solution. Can you provide me credentials to test the content of the solution. you can mail me them at [v-prasadboke@microsoft.com](mailto:v-prasadboke@microsoft.com)

Sent details to specified email

**Comment by v-prasadboke** [comment]
Receieved

**Comment by v-prasadboke** [comment]
Hello @nitsan-tzur, Can you share working images of function app. I tried deploying it in 3 different workspaces and subscriptions but every time I'm facing an error reasons being location and Dynamic VM's

**Comment by nitsan-tzur** [comment]
> Hello @nitsan-tzur, Can you share working images of function app. I tried deploying it in 3 different workspaces and subscriptions but every time I'm facing an error reasons being location and Dynamic VM's

Hi, This deployment template settings worked for me:
image (2)

Please note that the [aka.ms](https://aka.ms/sentinel-WithSecureElementsViaFunction-functionapp) URL seems to be wrong. It does not lead to the package

**Comment by v-prasadboke** [comment]
Hello @nitsan-tzur, 
You attached a screenshot of deployment page which had credentials. Which may had lead to unwanted activities. 
I have deleted the screenshot. 

Can you share working screenshot of function app. ( Output/Result )

**Comment by nitsan-tzur** [comment]
> Hello @nitsan-tzur, You attached a screenshot of deployment page which had credentials. Which may had lead to unwanted activities. I have deleted the screenshot.
> 
> Can you share working screenshot of function app. ( Output/Result )

Hi @v-prasadboke,
Thanks for that. Have you noticed the aka.ms link leads to 404: https://aka.ms/sentinel-CyberArkAudit-functionapp?

**Comment by v-prasadboke** [comment]
Reason being the hyperlink we have used in Shortlink leads to master branch. And this Solution is not in master branch yet. 
That is the reason. 

Thanks for sharing the screenshots.

**Comment by v-prasadboke** [comment]
Hello @nitsan-tzur, any reasons behind changing the Data connector folder structure

**Comment by v-prasadboke** [comment]
Hello @nitsan-tzur, Sorry for the late reply. 
I'll have to check once on the folder structure.

**Comment by nitsan-tzur** [comment]
> Hello @nitsan-tzur, Sorry for the late reply. I'll have to check once on the folder structure.

Hi @v-prasadboke 
Thanks for looking into it. Any update on this check?

**Comment by v-prasadboke** [comment]
Hello @nitsan-tzur, And sorry for the late reply. 

Function app's Folder structure is incorrect. We are working on it and I'll push the changes by tomorrow 27 March, 2024

**Comment by nitsan-tzur** [comment]
Hi @v-prasadboke ,
Thanks for the update. Will the PR be approved with this change?

**Comment by v-prasadboke** [comment]
Hello @nitsan-tzur, 
Folder structure corrected. But function.json is missing from the folder. 
Please add it. 
You can refer this one. Solutions/Box/Data Connectors/AzureFunctionBox/function.json
It should be placed in Functionapp zip as well.

**Comment by nitsan-tzur** [comment]
> Hello @nitsan-tzur, Folder structure corrected. But function.json is missing from the folder. Please add it. You can refer this one. Solutions/Box/Data Connectors/AzureFunctionBox/function.json It should be placed in Functionapp zip as well.

Hi @v-prasadboke,
It seems like repackaging is breaking the deployment. Might be due to movement of `requirements.json`. Just tried to deploy the new zip and function is empty:
image

**Comment by v-prasadboke** [comment]
Have you added function.json file

**Comment by nitsan-tzur** [comment]
Yes. Same result with the function.json file

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by v-prasadboke** [comment]
image

image

I can see connector on my end

**Review by v-prasadboke — DISMISSED** [review]

**Review by v-prasadboke — DISMISSED** [review]

**Comment by nitsan-tzur** [comment]
Hi @v-prasadboke ,
Thanks for your review. 
What should be the permeant aka.ms link to the zip package in the deployment template?

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**